File: composeApp/src/androidMain/kotlin/com/monzon/logssample/logger/SentryAntiLog.kt
package com.monzon.logssample.logger

import io.github.aakira.napier.Antilog
import io.github.aakira.napier.LogLevel
import io.sentry.*
import io.sentry.protocol.Message

class SentryAntiLog(
    private val hub: IHub,
    private val minEventLevel: SentryLevel,
    private val minBreadcrumbLevel: SentryLevel
) : Antilog() {
    override fun performLog(
        priority: LogLevel,
        tag: String?,
        throwable: Throwable?,
        message: String?
    ) {
        if (message.isNullOrEmpty() && throwable == null) {
            return
        }

        val level = getSentryLevel(priority)
        val sentryMessage = Message().apply {
            this.message = message
        }
        captureEvent(
            logLevel = level,
            tag = tag,
            msg = sentryMessage,
            throwable = throwable
        )

        addBreadcrumb(
            logLevel = level,
            msg = sentryMessage,
            throwable = throwable
        )
    }


    private fun captureEvent(
        logLevel: SentryLevel,
        tag: String?,
        msg: Message,
        throwable: Throwable?
    ) {
        if (isLoggable(logLevel, minEventLevel)) {
            val sentryEvent = SentryEvent().apply {
                level = logLevel
                throwable?.let { setThrowable(it) }
                tag?.let {
                    setTag("NapierTag", it)
                }
                message = msg
                logger = "Napier"
            }
            hub.captureEvent(sentryEvent)
        }
    }

    private fun addBreadcrumb(
        logLevel: SentryLevel,
        msg: Message,
        throwable: Throwable?
    ) {
        // checks the breadcrumb level
        if (isLoggable(logLevel, minBreadcrumbLevel)) {
            val throwableMsg = throwable?.message
            val breadCrumb = when {
                msg.message != null -> Breadcrumb().apply {
                    level = logLevel
                    category = "Napier"
                    message = msg.formatted ?: msg.message
                }

                throwableMsg != null -> Breadcrumb.error(throwableMsg).apply {
                    category = "exception"
                }

                else -> null
            }
            breadCrumb?.let { hub.addBreadcrumb(it) }
        }
    }

    private fun isLoggable(
        level: SentryLevel,
        minLevel: SentryLevel
    ): Boolean = level.ordinal >= minLevel.ordinal

    private fun getSentryLevel(logLevel: LogLevel): SentryLevel {
        return when (logLevel) {
            LogLevel.ASSERT -> SentryLevel.FATAL
            LogLevel.ERROR -> SentryLevel.ERROR
            LogLevel.WARNING -> SentryLevel.WARNING
            LogLevel.INFO -> SentryLevel.INFO
            LogLevel.DEBUG -> SentryLevel.DEBUG
            LogLevel.VERBOSE -> SentryLevel.DEBUG
            else -> SentryLevel.DEBUG
        }
    }

}

File: iosApp/iosApp/Logger/SentryAntiLog.swift
//
//  SentryAntiLog.swift
//  iosApp
//
//  Created by Carlos Monzon on 10/8/2024.
//  Copyright © 2024 orgName. All rights reserved.
//

import Foundation
import ComposeApp
import Sentry

class SentryAntiLog: NapierAntilog {
    
    let minEventLevel: SentryLevel
    let minBreadcumbLevel: SentryLevel
    
    init(minEventLevel: SentryLevel, minBreadcumbLevel: SentryLevel) {
        self.minEventLevel = minEventLevel
        self.minBreadcumbLevel = minBreadcumbLevel
    }
    
    override func performLog(priority: NapierLogLevel, tag: String?, throwable: KotlinThrowable?, message: String?) {
        guard message != nil || throwable != nil else { return }
        let level = getSentryLevel(level: priority)
        captureEvent(logLevel: level, tag: tag, msg: message ?? "", error: throwable?.asError())
        addBreadcrumb(logLevel: level, tag: tag, message: message, error: throwable?.asError())
    }
    
    private func addBreadcrumb(logLevel: SentryLevel, tag: String?, message: String?, error: Error?) {
        if (isLoggable(level: logLevel, minLevel: minBreadcumbLevel)) {
            let breadcrumb = Breadcrumb(level: logLevel, category: "Napier")
            breadcrumb.message = message
            if let error {
                breadcrumb.message = error.localizedDescription
            } else {
                breadcrumb.message = message
            }
            SentrySDK.addBreadcrumb(breadcrumb)
        }
    }
    
    private func captureEvent(
            logLevel: SentryLevel,
            tag: String?,
            msg: String,
            error: Error?
    ) {
        if (isLoggable(level: logLevel, minLevel: minEventLevel)) {
            let event = Event()
            event.level = logLevel
            event.error = error
            if let tag {
                event.tags = ["NapierTag": tag]
            }
            event.message = SentryMessage(formatted: msg)
            SentrySDK.capture(event: event)
        }
    }
    
    private func isLoggable(level: SentryLevel, minLevel: SentryLevel) -> Bool {
        level.rawValue >= minLevel.rawValue
    }
    
    private func getSentryLevel(level: NapierLogLevel) -> SentryLevel {
        switch level {
        case .debug:
            return SentryLevel.debug
        case .info:
            return SentryLevel.info
        case .warning:
            return SentryLevel.warning
        case .error:
            return SentryLevel.error
        case .assert:
            return SentryLevel.fatal
        default:
            return SentryLevel.debug
        
        }
    }
}
